Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -77,149 +78,207 @@ extension NIOClientTCPBootstrap {
                 if configuration.proxy != nil {
-                    bootstrap = NIOClientTCPBootstrap(tsBootstrap, tls: NIOInsecureNoTLS())
+                    bootstrap = eventLoop.makeSucceededFuture(NIOClientTCPBootstrap(tsBootstrap,
+                                                                                    tls: NIOInsecureNoTLS()))
                 } else {
                     // create NIOClientTCPBootstrap with NIOTS TLS provider
                     let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
                     let parameters = tlsConfiguration.getNWProtocolTLSOptions()
                     let tlsProvider = NIOTSClientTLSProvider(tlsOptions: parameters)
-                    bootstrap = NIOClientTCPBootstrap(tsBootstrap, tls: tlsProvider)
+                    bootstrap = eventLoop.makeSucceededFuture(NIOClientTCPBootstrap(tsBootstrap, tls: tlsProvider))
                 }
             } else if let clientBootstrap = ClientBootstrap(validatingGroup: eventLoop) {
-                bootstrap = try clientBootstrap.makeClientTCPBootstrap(host: host, requiresTLS: requiresTLS, configuration: configuration)
+                bootstrap = sslContextCache.sslContext(tlsConfiguration: configuration.tlsConfiguration ?? .forClient(),
+                                                       eventLoop: eventLoop,
+                                                       logger: logger)
+                    .flatMapThrowing { sslContext in
+                        try clientBootstrap.makeClientTCPBootstrap(host: host,
+                                                                   sslContext: sslContext,
+                                                                   useProxy: configuration.proxy != nil)
+                    }
             } else {
                 preconditionFailure("Cannot create bootstrap for the supplied EventLoop")
             }
         #else
             if let clientBootstrap = ClientBootstrap(validatingGroup: eventLoop) {
-                bootstrap = try clientBootstrap.makeClientTCPBootstrap(host: host, requiresTLS: requiresTLS, configuration: configuration)
+                bootstrap = sslContextCache.sslContext(tlsConfiguration: configuration.tlsConfiguration ?? .forClient(),
+                                                       eventLoop: eventLoop,
+                                                       logger: logger)
+                    .flatMapThrowing { sslContext in
+                        try clientBootstrap.makeClientTCPBootstrap(host: host,
+                                                                   sslContext: sslContext,
+                                                                   useProxy: configuration.proxy != nil)
+                    }
             } else {
                 preconditionFailure("Cannot create bootstrap for the supplied EventLoop")
             }
```

Comment:
Since this is very much the same code as L101-111, do you think we can reorganize this into something like this:

```swift
func actuallyCreateBootstrap() -> EventLoopFuture<NIOClientTCPBootstrap> {
    #if canImport(Network)
        if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
            return nioTSBootstrap
        }
        // if we can not create a nioTSBootstrap just fallthrough
    #endif

    guard let clientBootstrap = ClientBootstrap(validatingGroup: eventLoop) else {
        preconditionFailure("Cannot create bootstrap for the supplied EventLoop")
    }

    return nioBootstrap
}
```